bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/discovery/SerialNumberHelper.java

```diff
@@ -0,0 +1,92 @@
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.http.HttpStatus;
+import org.openhab.binding.hdpowerview.internal.dto.UserData;
+import org.openhab.binding.hdpowerview.internal.dto.gen3.Info;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * This helper class fetches the hub serial number so it can be used as the representation property
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = SerialNumberHelper.class)
+public class SerialNumberHelper {
+
+    public enum ApiVersion {
+        V1,
+        V3;
+    }
+
+    private static final String API_V1_URL_FORMAT = "http://%s/api/userdata";
+    private static final String API_V3_URL_FORMAT = "http://%s/gateway/info";
+
+    private final Gson gsonParser = new Gson();
+    private final HttpClientFactory httpClientFactory;
+
+    @Activate
+    public SerialNumberHelper(@Reference HttpClientFactory httpClientFactory) {
+        this.httpClientFactory = httpClientFactory;
+    }
+
+    /**
+     * Get the serial number of the hub / gateway on the given IPv4 address
+     *
+     * @param ipAddress a dotted ipv4 address
+     * @param apiVersion the hub / gateway api version used
+     * @return the serial number
+     */
+    public String getSerialNumber(String ipAddress, ApiVersion apiVersion) {
+        try {
+            HttpClient httpClient = httpClientFactory.getCommonHttpClient();
+            String uri = String.format(ApiVersion.V1 == apiVersion ? API_V1_URL_FORMAT : API_V3_URL_FORMAT, ipAddress);
+            ContentResponse content = httpClient.GET(uri);
+            if (HttpStatus.OK_200 == content.getStatus()) {
+                String json = content.getContentAsString();
+                switch (apiVersion) {
+                    case V1:
+                        UserData userData = gsonParser.fromJson(json, UserData.class);
+                        if (userData != null && userData.serialNumber instanceof String serialNumber) {
+                            return serialNumber;
+                        }
+                        break;
+
+                    case V3:
+                        Info info = gsonParser.fromJson(json, Info.class);
+                        if (info != null && info.getSerialNumber() instanceof String serialNumber) {
+                            return serialNumber;
+                        }
+                        break;
+                }
+            }
+        } catch (JsonSyntaxException | InterruptedException | ExecutionException | TimeoutException e) {
+        }
+        // fall back to ip address
+        return ipAddress;
```

Comment:
This seems a bit strange, I guess a caller would expect to get a serial number, not an IP address. Perhaps it would be better to throw an exception if this isn't possible, so that the caller itself can contain the logic for falling back to IP address when that makes sense?

Reply:
Falling back to ip address is basically falling back to the current behaviour. Which was not completely bad. Which is the reason why I chose to do that.

Reply:
The comment was not about the fallback itself, but rather the way it's implemented: I would expect a method called `getSerialNumber` to return a serial number or throw an exception if something went wrong. The discovery classes could then catch this exception and fall back if desired.

But thinking about it again, does it make sense? I create a discovered Thing with property detected serial number. I then trigger another discovery and the API call then fails. So it falls back to IP address, which doesn't match my serial number, so it will be discovered again anyway? That reminds me to double-check that all Thing handlers will initialize the `serialNumber` property, maybe you already verified that?

Reply:
>  expect a method called getSerialNumber to return a serial number or throw an exception if something went wrong

Ok. 

> double-check that all Thing handlers will initialize the serialNumber property, maybe you already verified that?

Yes.